### PR TITLE
[#IC-49] Porting SpecialServices to new ProcessMessage flow

### PR DESCRIPTION
**Pull request by BurnedMarshal**
#### List of Changes

Upgrade `io-functions-commons` dependency for SpecialServices
CreateMessage flow now have `serviceCategory` to handle Standard and Special service preferences check.
Only admin can create and edit `SpecialService` properties.

#### Motivation and Context

The `io-functions-services` should handle Standard and Special services. This PR upgrade specifications and models to do that.
When a message is created for a Special Service to evaluate if INBOX is enabled for a User, Activation model is used to retrieve Special Service activation.

#### How Has This Been Tested?

unit test

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: ProcessMessage/handler.ts [hunk 1]
```diff
@@ -384,6 +389,49 @@ export const getProcessMessageHandler = ({
 
               return result;
             }),
+            TE.chain(_ => {
```
Comment (balanza) [inline ProcessMessage/handler.ts:392]: Consider move this sub-pipe into a dedicated function to increase redability and keep the main pipe at the same level of abstraction

File: UpdateService/handler.ts [hunk 2]
```diff
@@ -137,10 +141,23 @@ const updateServiceTask = (
...
+          service_metadata: SpecialServiceMetadata.is(
+            retrievedService.service_metadata
+          )
+            ? ({
+                ...servicePayload.service_metadata,
+                category: SpecialServiceCategoryEnum.SPECIAL,
+                custom_special_flow:
+                  retrievedService.service_metadata.custom_special_flow,
+                token_name: adb2cTokenName
+              } as SpecialServiceMetadata)
```
Comment (balanza) [inline UpdateService/handler.ts:154]: Why the need for manual casting? I'm afraid it may hide some type inconsistency

File: openapi/index.yaml [hunk 3]
```diff
@@ -756,7 +756,7 @@ definitions:
           created_at:
             $ref: '#/definitions/Timestamp'
           content:
-            $ref: '#/definitions/MessageContent'
+            $ref: '#/definitions/NewMessageContent'
```
Comment (balanza) [inline openapi/index.yaml:759]: I see this change to come from template rendering, I just wonder why it happened.

File: package.json [hunk 4]
```diff
@@ -20,7 +20,7 @@
     "generate": "npm-run-all generate:*",
     "generate:definitions": "rimraf ./generated/definitions && shx mkdir -p ./generated/definitions && gen-api-models --api-spec ./openapi/index.yaml --no-strict --out-dir ./generated/definitions",
     "generate:api-notifications": "rimraf ./generated/notifications && shx mkdir -p ./generated/notifications && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-backend/master/api_notifications.yaml --out-dir ./generated/notifications --response-decoders --request-types",
-    "generate:api-admin": "rimraf generated/api-admin && shx mkdir -p generated/api-admin && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-functions-admin/master/openapi/index.yaml --no-strict --out-dir generated/api-admin --request-types --response-decoders --client",
+    "generate:api-admin": "rimraf generated/api-admin && shx mkdir -p generated/api-admin && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-functions-admin/IC-40-special-service-model-upgrade/openapi/index.yaml --no-strict --out-dir generated/api-admin --request-types --response-decoders --client",
```
Comment (balanza) [inline package.json:23]: We won't merge a specific branch. Consider:
* generate code from `master` (as it used to be)
* generate code from a specific version (good)
* import sdk package (better)

**Review by balanza — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (BurnedMarshal) [reply package.json:23]: You have right. This should be changed to `master` when the pagopa/io-functions-admin#172 is merged.
If we want to move to sdk import version I think is better do the refactory on another PR.

File: openapi/index.yaml (hunk 3 above)
Comment (BurnedMarshal) [reply openapi/index.yaml:759]: This change comes from [here](https://github.com/pagopa/io-functions-commons/pull/218/files#diff-390ee41a01068fe2e94192fa8caf0b01f8ad5e5a71545ae89874cf453348b334R267) released with `v21.7.3` of `io-functions-commons`

File: UpdateService/handler.ts (hunk 2 above)
Comment (BurnedMarshal) [reply UpdateService/handler.ts]: Good catch. All remove all casting. Fixed in 48003b8

File: ProcessMessage/handler.ts (hunk 1 above)
Comment (gunzip) [reply ProcessMessage/handler.ts:392]: I'd also add a comment on what's happening

Comment (BurnedMarshal) [reply ProcessMessage/handler.ts]: Fixed in 8a9cfb5

Comment (BurnedMarshal) [reply ProcessMessage/handler.ts]: I've also added a comment in b2e90ed

**Comment by BurnedMarshal** [comment]
> override with what? (why?) it's not very clear form the comment how this works

Sorry, I try to explain it better

**Comment by BurnedMarshal** [comment]
> > override with what? (why?) it's not very clear form the comment how this works
> 
> Sorry, I try to explain it better

@gunzip I've fixed the comments 2ca3a1a. Tell me if now is more clear.

File: ProcessMessage/handler.ts [hunk 5]
```diff
@@ -384,6 +452,26 @@ export const getProcessMessageHandler = ({
...
               return result;
             }),
+            TE.chain(blockedInboxOrChannel => {
+              if (
+                createdMessageEvent.senderMetadata.serviceCategory ===
+                SpecialServiceCategoryEnum.SPECIAL
+              ) {
+                // If the service has category equals to SPECIAL, the INBOX value in blockedInboxOrChannel depends from the Activation
+                // value. If the SPECIAL service is ACTIVE (exists an activation for couple user/service with ACTIVE status)
+                // we remove INBOX from blockedInboxOrChannel, otherwise we add the INBOX value.
```
Comment (gunzip) [inline ProcessMessage/handler.ts:462]:
```suggestion
```

File: ProcessMessage/handler.ts [hunk 6]
```diff
@@ -384,6 +452,26 @@ export const getProcessMessageHandler = ({
...
+                return getActivationForSpecialServices(lActivation)({
```
Comment (gunzip) [inline ProcessMessage/handler.ts:463]:
```suggestion
                return getBlockedInboxesForSpecialService(lActivation)({
```

File: ProcessMessage/handler.ts [hunk 7]
```diff
@@ -384,6 +452,26 @@ export const getProcessMessageHandler = ({
...
+                  blockedInboxOrChannel,
+                  context,
+                  createdMessageEvent,
```
Comment (gunzip) [inline ProcessMessage/handler.ts:466]: it looks like you only need the `senderServiceId`

File: ProcessMessage/handler.ts [hunk 8]
```diff
@@ -183,6 +186,69 @@ const getServicePreferenceValueOrError = (
...
+}) => TaskEither<
+  ReadonlyArray<BlockedInboxOrChannelEnum>,
+  ReadonlyArray<BlockedInboxOrChannelEnum>
+>;
+
+/**
+ * Returns a new blockedInboxOrChannel value for a SPECIAL service related to a User.
+ * When a service has category SPECIAL the INBOX value depends from the activation status
+ * for the couple user/service.
+ * A missing Activation is equal to an INACTIVE one.
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply ProcessMessage/handler.ts]: Thanks a lot :D

File: ProcessMessage/handler.ts [hunk 9]
```diff
@@ -183,6 +186,69 @@ const getServicePreferenceValueOrError = (
...
+ *
+ * @param lActivation
+ * @returns
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply ProcessMessage/handler.ts:213]: If you think that is more clear we can return a `TaskEither<never, ReadonlyArray<BlockedInboxOrChannelEnum>>` changing the last part of the chain like this:
```typescript
TE.chainW(
  TE.fromPredicate(hasActiveActivation => hasActiveActivation, identity)
),
TE.fold(
  () =>
    TE.of(
      blockedInboxOrChannel.includes(BlockedInboxOrChannelEnum.INBOX)
        ? blockedInboxOrChannel
        : [...blockedInboxOrChannel, BlockedInboxOrChannelEnum.INBOX]
    ),
  () =>
    TE.of(
      blockedInboxOrChannel.filter(
        el => el !== BlockedInboxOrChannelEnum.INBOX
      )
   )
)
```

File: ProcessMessage/handler.ts (hunk 7 above)
Comment (BurnedMarshal) [reply ProcessMessage/handler.ts]: Fixed in 0b1169e

File: ProcessMessage/handler.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply ProcessMessage/handler.ts:217]: a comment was added in 0b1169e

[the remark this answers is not in the document]
Comment (gunzip) [reply ProcessMessage/handler.ts:217]: it depends on the usage, if you only use one side returning a `Task` would be better (or using `never`).

**Review by gquadrati — APPROVED** [review]
LGTM

File: ProcessMessage/handler.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply ProcessMessage/handler.ts:210]: `TaskEither` was replaced by a `Task` in 06a1b5c

File: package.json (hunk 4 above)
Comment (BurnedMarshal) [reply package.json]: Fixed in 988385e

File: UpdateService/handler.ts [hunk 10]
```diff
@@ -137,10 +140,22 @@ const updateServiceTask = (
           ...retrievedService,
           ...servicePayload,
           service_id: serviceId,
-          service_metadata: {
-            ...servicePayload.service_metadata,
-            token_name: adb2cTokenName
-          }
+          // Only Admins can change service category and custom_special_flow name
```
Comment (balanza) [inline UpdateService/handler.ts:143]: It's not clear to me how "being admin" is ensured in this flow

Comment (BurnedMarshal) [reply UpdateService/handler.ts:143]: Good point. I've tried to extend the comment in f2482cb

**Pull request merged**